### PR TITLE
[crypto/bn] constant-time modular inversion

### PR DESCRIPTION
**Pull request by bbbrumley**
This is joint work with @sohhas. We did some work in NSS for CT inversion last year, and this patch is roughly half porting that work, and half new code due to differences in inversion strategies between the libraries.

The diff is not really useful to review here. Take out the `BN_gcd` function, and the rest of the code is almost all new.

In `master`, there are essentially three algorithms to compute the same thing using different ways; none are completely constant time (see e.g. [Sec. 6.5](https://www.usenix.org/conference/usenixsecurity20/presentation/weiser)).

The new code also has three algorithms:

1. CT inversion with an odd modulus
2. CT inversion with a power-of-two modulus
3. CT inversion with a power-of-two modulus where stuff fits in words

For generic even modulus, there's some CRT stuff to compute the result with (1), either (2) or (3), followed by some short math. A short discussion on that follows. For random even modulus, the typical flow is (1,3,3). Outside the artificial KATs added in this PR, I never hit the (1,2,2) case in `make test`. So that is why (3) exists, even if it is functionally redundant given (2): it is faster and cleaner.

For (EC)DSA stuff and in general ECC, all that matters is (1) above.

RSA-related inversions are really the only other concern here. For a "random looking" even modulus, the flow through this code will be mostly (1) with a large modulus and (3) with an overwhelmingly small modulus. It means that, performance wise, there shouldn't be much of difference between an odd modulus and a random even modulus.

For the even modulus case, this code assume the power-of-two (i.e., the output of "count trailing zeros" given the modulus) is public. I claim this is OK in the context of RSA in the (rare) case of using any secret composite moduli; consider these two scenarios.
1. I am a side-channel capable attacker. I use a lot of measurement effort and somehow learn the (fixed) power-of-two, perhaps with repeated measurements and profiling. I use this information in some super special attack.
2. I am not a side-channel capable attacker. I guess the (fixed) power-of-two, and run the same super special attack with that. For a b-bit modulus, there are only b powers to check. So at most I run b instances of my attack, and the probability of the correct b is negligible in b.

This is the end of me whining that it's hard to protect the power-of-two.

Edit: In 537f242bb9bc027a7b4236d7a2be2e7091c1e405 we reduced the power-of-two leakage to the word position of the power-of-two.

We generated the unit tests deterministically with a python script; we can provide that upon request.

Tagging @cpereida for (unofficial) review.

### CODE REVIEW

File: crypto/bn/bn_gcd.c [hunk 1]
```diff
@@ -525,9 +292,29 @@ BIGNUM *BN_mod_inverse(BIGNUM *in,
...
+        if ((n_pos = BN_CTX_get(ctx)) == NULL
+            || BN_copy(n_pos, n) == NULL)
+            goto err;
+        n_pos->neg = 0;
+    }
+
+    /* ensure a is non-negative */
+    if (BN_is_negative(a) && !BN_is_zero(n)) {
+        if ((a_pos = BN_CTX_get(ctx)) == NULL
+            || !BN_nnmod(a_pos, a, n, ctx))
```
Comment (romen) [inline crypto/bn/bn_gcd.c:308]: Should this conditionally use n_pos?

**Comment by paulidale** [comment]
I don't think it will be a surprise that this is post 3.0 :)

The test generating script would be interesting I think.  Perhaps in the openssl/tools repository?  We've other [test generation](https://github.com/openssl/tools/blob/master/statistics/bn_rand_range.py) there in Python.

File: crypto/bn/bn_gcd.c (hunk 1 above)
Comment (bbbrumley) [reply crypto/bn/bn_gcd.c:308]: I thought about it at the time, but it seems in `BN_nnmod` the sign of the modulus is a don't care. I'll double check, though.

**Comment by t8m** [comment]
The CI failures on Windows builds are relevant.

**Comment by bbbrumley** [comment]
> The CI failures on Windows builds are relevant.

I don't know how to make those false positives go away without [stuff like this](https://gitlab.com/nisec/ecckiila/-/blob/master/tpl/scalar_mul.c#L211) ... any tips?

(at least, I am pretty sure they are false positives ...)

**Comment by paulidale** [comment]
That kind of pragma mess is the best way I think.  We don't have BN_LONG to cast to.

The only other thing is to emulate the 2s complement negation `-x` <=> `1 + ~x` which is worse.  I'm pretty sure we rely on 2s complement arithmetic in various places.

**Comment by paulidale** [comment]
I can't believe that that fix worked....

**Comment by bbbrumley** [comment]
> I can't believe that that fix worked....

It sounds [ridiculous](https://devblogs.microsoft.com/oldnewthing/20210518-00/?p=105225), doesn't it! Untested, so let's see what the CI says ..

File: crypto/bn/bn_gcd.c (hunk 1 above)
Comment (bbbrumley) [reply crypto/bn/bn_gcd.c]: added as part of bad6659b1ef56c37a0887df82b0ccbdc901bc037

**Comment by bbbrumley** [comment]
I forgot to say -- this PR makes the function oblivious to the `BN_FLG_CONSTTIME` flag.

(Which, I _think_ by my count, means only modular exp left!)

**Comment by romen** [comment]
I opened #16232 to discuss within OTC the option to remove support for `BN_mod_*` operations with negative moduli; in the meantime should we add KAT for negative moduli? Otherwise the negative modulus codepath goes untested, right?

**Comment by bbbrumley** [comment]
> should we add KAT for negative moduli? Otherwise the negative modulus codepath goes untested, right?

Good point! Can't believe I missed that. Fixed in 0274ffd6a208174c320550a8524a32d88dfbb1f8

Maybe I should sanity check with `master` just in case.

**Comment by davidben** [comment]
> For the even modulus case, this code assume the power-of-two (i.e., the output of "count trailing zeros" given the modulus) is public.

Other than setting up Montgomery contexts (which only inverts a single word), the one place I can think of with a secret even modulus is in RSA key generation, to compute `d = e^1 (mod totient)`.

You're probably right that the output of count trailing zeros can be assumed public here, but this is also not a performance-sensitive part of RSA keygen. The primality tests dominate by far. Rather than worry about this, I adapted the binary GCD algorithm. It's not particularly fast, but it doesn't matter for this.
https://boringssl.googlesource.com/boringssl/+/refs/heads/master/crypto/fipsmodule/bn/gcd_extra.c#169

Unless I've forgotten one, I believe all other inversions in cryptography use either a public prime modulus, or a public near-prime modulus (RSA n). The former we do with FLT, and the latter we do with a blinding trick:

```
// May fail if n has too many factors, but n = p*q, so it is as likely as
// accidentally stumbling on p or q.
mod_inverse(a, n):
  r = random(1, n)           // May leak n, but n is public. Does not leak secret r.
  b = mod_mul(a, r, n)       // Montgomery reduction, so no leak of a or r.
  inv = leaky_mod_inv(b, n)  // Leaks b, but b can be public. Note inv = a^-1 * r^-1 (mod n)
  return mod_mul(inv, r, n)  // Montgomery reduction, so no leak of r
```

**Comment by bbbrumley** [comment]
> You're probably right that the output of count trailing zeros can be assumed public here, but this is also not a performance-sensitive part of RSA keygen

@sohhas and I added some hardening in 537f242bb9bc027a7b4236d7a2be2e7091c1e405

It now leaks the position of the least significant non-zero word. Since the code inherently diverges based on that position, it doesn't make sense to try any harder than that (with this approach).

>  the one place I can think of with a secret even modulus is in RSA key generation

Yes computations modulo the totient is also all we had in mind, too. So the above will leak fixed information for roughly one key in every `2**BN_BITS2` keys, which is not significant IMO. (Again, an attacker can just assume that and run the attack with no side-channel information.)

**Comment by davidben** [comment]
> It now leaks the position of the least significant non-zero word.

Yeah, that'll require addressing the ever-present #6640. I can report it is indeed possible to solve that problem.

But I don't think that commit addresses `idx_o`. Even computed constant-time, it's leaked in all the other operations. `BN_rshift`, for instance, will access memory differently depending on the amount you shift by. (We have a `bn_rshift_secret_shift` that conditionally shifts by powers of two for the couple of places this matters. Both, I believe, in RSA key generation.) `bn_mod_inverse_pow2 ` also appears to treat the modulus as public. And, of course, by shifting bits away, you change the size of the input and, in turn, means #6640 kicks in further. 

(One of these days I should write up what it took to get to constant-time RSA keygen. It's doable but takes quite a lot of bashing on the problem. There are a tons of operations unique to RSA keygen.)

**Comment by bbbrumley** [comment]
> `bn_mod_inverse_pow2 ` also appears to treat the modulus as public

Thanks! Made some adjustments in 1077f3137eed8f7ba7255fdb9ab9deb73d2ec160, considering the overwhelming codepath through `bn_mod_inverse_pow2` is to take the early exit to `bn_mod_inverse_pow2_word`.

**Comment by bbbrumley** [comment]
The Windows CI is failing after 1077f31 and I have no idea :man_shrugging: at this point I'm just trying things and querying the CI. Could use some help since I'm a linux-only dude.

I thought it was undefined behavior because of the shift, but it seems that's not the case or a52f5d20dca8c2068d805136f161b8d9410ff150 would've fixed it.

Any ideas?

**Comment by bbbrumley** [comment]
Mkay I'm a bit baffled that `BN_MASK2` isn't defined with `ULL`, but anyway f2cc5d5119508df5a33832375bf180062ce37dc7 fixes it.

**Comment by davidben** [comment]
It seems this still treats the power of two as public:

* It leaks whether we took `bn_mod_inverse_pow2` or `bn_mod_inverse_pow2_word`
* `bn_mod_inverse_pow2` calls `BN_set_bit(t0, e)` which touches a different word depending on `e`. Likewise with `BN_mask_bits`.
* `bn_mod_inverse_pow2` counts up to `e`

FWIW, I don't disagree with your assessment that leaking the power of two is probably fine. If the strategy is to leak `e` on grounds that it's probably benign, then leak `e`. I believe it's possible to avoid it (the link I gave above), but it does seem this particular algorithm is resisting attempts to avoid it.

(There may even be cases you *want* to leak `e`, if `e` is public and it's perf-sensitive. For BoringSSL, we have several different modular inversion functions, since different places have different secrecy and performance requirements, and different assumptions on the inputs. Some are tailored to a particular public EC prime. Some work for arbitrary public primes. Some assume a secret modulus, some don't. One specifically inverts mod 2^64. And so on. To go down that route, I think the first step is to take inventory of the places where you need to compute modular inversions.)

**Comment by bbbrumley** [comment]
> It seems this still treats the power of two as public

Not within the word :) That's what I meant by "It now leaks the position of the least significant non-zero word". I hope that's clear.

>     * It leaks whether we took `bn_mod_inverse_pow2` or `bn_mod_inverse_pow2_word`
>     * `bn_mod_inverse_pow2` calls `BN_set_bit(t0, e)` which touches a different word depending on `e`. Likewise with `BN_mask_bits`.

Yes -- and that's "up to word".

>     * `bn_mod_inverse_pow2` counts up to `e`

Indeed; for random inputs, that counting will happen with probability `2**-BN_BITS2`.

**Comment by bernd-edlinger** [comment]
Hello @bbbrumley nice work!
This PR is marked as post3.0.0 however, it seems to be simple to back-port to the 1.1.1 branch.
I would like to ask if I may add your PR to my openssl-111-features branch, see #11900

**Comment by bbbrumley** [comment]
Hey Bernd,

Thanks for the interest!

Hmm yea in general, I don't see why not.

I still think I'm missing a little due diligence on this one. I didn't do any performance analysis at all.

Unrelated: much obliged for [this patch of yours](https://repo.or.cz/alpine.git/commit/529d58309bc7ec042b668a90817eabe4522507e0)! Everyone on this thread is witnessing a tiny miracle: two (al)pine S/MIME users meeting in the wild.

**Comment by bbbrumley** [comment]
Right so ... after mulling this one over for some time ... I've decided to close it.

I think @davidben is right.

At a high level, this PR

1. Removes some non-CT inversion code that leaks in certain known small ways that nobody knows how to exploit.
2. Replaces it with some non-CT inversion code that leaks in certain known small ways that nobody knows how to exploit.

No matter how much explaining away I try to do about how this PR leaks in better ways than the code it is replacing,
1. There's no guarantee that's true -- I am some random dude on the Interwebs.
2. There's certainly no guarantee that will hold moving forward, as more features / use cases get added to OpenSSL.

In summary, the code this PR attempted to replace was a crutch at the time, a very long time ago (2007ish?). I feel it is unwise to trade one crutch for another. At this point, do it properly, or don't do it at all.

Anyway, it was fun to implement :) I felt like a PhD student again, for a very brief moment. I'll go back to something incredibly boring like grant writing now.

**Pull request closed without merging**